File: parse_existing_data.py
#!/usr/bin/env python3
import os
import sqlite3
import csv
from datetime import datetime
from pathlib import Path

# Database path
DB_PATH = "stock_analysis_webapp/database/market_data.db"
DATA_DIR = "/Users/thiags/Downloads/data 2/daily/us"

# Exchange mappings
EXCHANGES = {
    'nasdaq stocks': 'NASDAQ',
    'nyse stocks': 'NYSE', 
    'nysemkt stocks': 'NYSEMKT',
    'nasdaq etfs': 'NASDAQ',
    'nyse etfs': 'NYSE',
    'nysemkt etfs': 'NYSEMKT'
}

def parse_csv_line(line):
    """Parse a CSV line from the data file"""
    parts = line.strip().split(',')
    if len(parts) < 10:
        return None
    
    ticker, period, date, time, open_price, high, low, close, volume, openint = parts
    
    # Skip header or invalid lines
    if ticker == '<TICKER>' or not ticker or not date:
        return None
    
    try:
        # Parse date (format: YYYYMMDD)
        year = date[:4]
        month = date[4:6]
        day = date[6:8]
        formatted_date = f"{year}-{month}-{day}"
        
        return {
            'symbol': ticker.replace('.US', '').replace('.us', ''),
            'date': formatted_date,
            'open': float(open_price),
            'high': float(high),
            'low': float(low),
            'close': float(close),
            'volume': float(volume)
        }
    except (ValueError, IndexError):
        return None

def process_data_file(file_path, exchange, db_conn):
    """Process a single data file"""
    processed = 0
    errors = 0
    
    try:
        with open(file_path, 'r', encoding='utf-8') as f:
            lines = f.readlines()
        
        # Skip header line
        for line in lines[1:]:
            parsed = parse_csv_line(line)
            
            if not parsed:
                errors += 1
                continue
            
            # Insert into database
            cursor = db_conn.cursor()
            try:
                cursor.execute("""
                    INSERT OR REPLACE INTO historical_prices 
                    (symbol, date, open, high, low, close, volume) 
                    VALUES (?, ?, ?, ?, ?, ?, ?)
                """, (
                    parsed['symbol'],
                    parsed['date'],
                    parsed['open'],
                    parsed['high'],
                    parsed['low'],
                    parsed['close'],
                    parsed['volume']
                ))
                processed += 1
            except sqlite3.Error as e:
                errors += 1
                print(f"❌ DB Error for {parsed['symbol']}: {e}")
            
            cursor.close()
        
    except Exception as e:
        print(f"❌ Error reading {file_path}: {e}")
        errors += 1
    
    return processed, errors

def process_directory(dir_path, exchange, db_conn):
    """Process all files in a directory recursively"""
    total_processed = 0
    total_errors = 0
    
    try:
        items = os.listdir(dir_path)
        
        for item in items:
            full_path = os.path.join(dir_path, item)
            
            if os.path.isdir(full_path):
                # Recursively process subdirectories
                processed, errors = process_directory(full_path, exchange, db_conn)
                total_processed += processed
                total_errors += errors
            elif item.endswith('.txt'):
                # Process data file
                processed, errors = process_data_file(full_path, exchange, db_conn)
                total_processed += processed
                total_errors += errors
                
                # Add symbol to symbols table
                symbol = item.replace('.txt', '').replace('.US', '').replace('.us', '')
                cursor = db_conn.cursor()
                
                try:
                    is_etf = 'etf' in exchange.lower()
                    cursor.execute("""
                        INSERT OR IGNORE INTO symbols (symbol, name, exchange) 
                        VALUES (?, ?, ?)
                    """, (symbol, symbol, exchange))
                except sqlite3.Error as e:
                    print(f"❌ Error adding symbol {symbol}: {e}")
                
                cursor.close()
    
    except Exception as e:
        print(f"❌ Error processing directory {dir_path}: {e}")
    
    return total_processed, total_errors

def update_data_freshness(db_conn):
    """Update data freshness table with summary information"""
    cursor = db_conn.cursor()
    
    try:
        # Get summary for each symbol
        cursor.execute("""
            SELECT symbol, COUNT(*) as count, MIN(date) as earliest, MAX(date) as latest
            FROM historical_prices 
            GROUP BY symbol
        """)
        
        rows = cursor.fetchall()
        
        # Update data freshness
        for row in rows:
            symbol, count, earliest, latest = row
            cursor.execute("""
                INSERT OR REPLACE INTO data_freshness 
                (symbol, last_updated, status, error_count) 
                VALUES (?, datetime('now'), 'active', 0)
            """, (symbol,))
        
        print(f"\n💾 Database updated with {len(rows)} symbols")
        
    except sqlite3.Error as e:
        print(f"❌ Error updating data freshness: {e}")
    
    cursor.close()

def main():
    """Main processing function"""
    print("🚀 Starting data import from existing files...")
    print(f"📁 Data directory: {DATA_DIR}")
    
    if not os.path.exists(DATA_DIR):
        print(f"❌ Data directory not found: {DATA_DIR}")
        return
    
    if not os.path.exists(DB_PATH):
        print(f"❌ Database not found: {DB_PATH}")
        return
    
    # Connect to database
    try:
        db_conn = sqlite3.connect(DB_PATH)
        print("✅ Connected to database")
    except sqlite3.Error as e:
        print(f"❌ Database connection error: {e}")
        return
    
    grand_total_processed = 0
    grand_total_errors = 0
    
    # Process each exchange directory
    for exchange_dir, exchange_name in EXCHANGES.items():
        exchange_path = os.path.join(DATA_DIR, exchange_dir)
        
        if os.path.exists(exchange_path):
            print(f"\n📊 Processing {exchange_dir} ({exchange_name})...")
            
            processed, errors = process_directory(exchange_path, exchange_name, db_conn)
            grand_total_processed += processed
            grand_total_errors += errors
            
            print(f"✅ {exchange_dir}: {processed:,} records, {errors} errors")
    
    print(f"\n🎉 Import Complete!")
    print(f"📊 Total records processed: {grand_total_processed:,}")
    print(f"❌ Total errors: {grand_total_errors}")
    
    # Update data freshness
    update_data_freshness(db_conn)
    
    # Commit and close
    db_conn.commit()
    db_conn.close()
    
    print("💾 Database import completed successfully!")

if __name__ == "__main__":
    main()

File: backend/parse_existing_data.js
const fs = require('fs');
const path = require('path');
const sqlite3 = require('sqlite3').verbose();

// Database path
const dbPath = path.join(__dirname, '../database/market_data.db');
const db = new sqlite3.Database(dbPath);

// Data directory path
const DATA_DIR = '/Users/thiags/Downloads/data 2/daily/us';

// Exchange mappings
const EXCHANGES = {
    'nasdaq stocks': 'NASDAQ',
    'nyse stocks': 'NYSE', 
    'nysemkt stocks': 'NYSEMKT',
    'nasdaq etfs': 'NASDAQ',
    'nyse etfs': 'NYSE',
    'nysemkt etfs': 'NYSEMKT'
};

// Parse CSV line
function parseCSVLine(line) {
    const parts = line.split(',');
    if (parts.length < 9) return null;
    
    const [ticker, period, date, time, open, high, low, close, volume] = parts;
    
    // Skip header or invalid lines
    if (ticker === '<TICKER>' || !ticker || !date) return null;
    
    // Parse date (format: YYYYMMDD)
    const year = date.substring(0, 4);
    const month = date.substring(4, 6);
    const day = date.substring(6, 8);
    const formattedDate = `${year}-${month}-${day}`;
    
    return {
        symbol: ticker.replace('.US', '').replace('.us', ''),
        date: formattedDate,
        open: parseFloat(open),
        high: parseFloat(high),
        low: parseFloat(low),
        close: parseFloat(close),
        volume: parseFloat(volume)
    };
}

// Process a single data file
async function processDataFile(filePath, exchange) {
    return new Promise((resolve, reject) => {
        fs.readFile(filePath, 'utf8', (err, data) => {
            if (err) {
                console.log(`❌ Error reading ${filePath}: ${err.message}`);
                resolve({ processed: 0, errors: 1 });
                return;
            }
            
            const lines = data.split('\n').filter(line => line.trim());
            let processed = 0;
            let errors = 0;
            
            // Skip header line
            for (let i = 1; i < lines.length; i++) {
                const line = lines[i];
                const parsed = parseCSVLine(line);
                
                if (!parsed) {
                    errors++;
                    continue;
                }
                
                // Insert into database
                const stmt = db.prepare(`
                    INSERT OR REPLACE INTO historical_prices 
                    (symbol, date, open, high, low, close, volume) 
                    VALUES (?, ?, ?, ?, ?, ?, ?)
                `);
                
                try {
                    stmt.run([
                        parsed.symbol,
                        parsed.date,
                        parsed.open,
                        parsed.high,
                        parsed.low,
                        parsed.close,
                        parsed.volume
                    ]);
                    processed++;
                } catch (dbErr) {
                    errors++;
                    console.log(`❌ DB Error for ${parsed.symbol}: ${dbErr.message}`);
                }
                
                stmt.finalize();
            }
            
            resolve({ processed, errors });
        });
    });
}

// Process all files in a directory
async function processDirectory(dirPath, exchange) {
    const items = fs.readdirSync(dirPath);
    let totalProcessed = 0;
    let totalErrors = 0;
    
    for (const item of items) {
        const fullPath = path.join(dirPath, item);
        const stat = fs.statSync(fullPath);
        
        if (stat.isDirectory()) {
            // Recursively process subdirectories
            const result = await processDirectory(fullPath, exchange);
            totalProcessed += result.processed;
            totalErrors += result.errors;
        } else if (item.endsWith('.txt')) {
            // Process data file
            const result = await processDataFile(fullPath, exchange);
            totalProcessed += result.processed;
            totalErrors += result.errors;
            
            // Add symbol to symbols table
            const symbol = item.replace('.txt', '').replace('.US', '').replace('.us', '');
            const stmt = db.prepare(`
                INSERT OR IGNORE INTO symbols (symbol, name, exchange, type) 
                VALUES (?, ?, ?, ?)
            `);
            
            try {
                const isETF = exchange.toLowerCase().includes('etf');
                stmt.run([symbol, symbol, exchange, isETF ? 'ETF' : 'STOCK']);
            } catch (err) {
                console.log(`❌ Error adding symbol ${symbol}: ${err.message}`);
            }
            stmt.finalize();
        }
    }
    
    return { processed: totalProcessed, errors: totalErrors };
}

// Main processing function
async function importAllData() {
    console.log('🚀 Starting data import from existing files...');
    console.log(`📁 Data directory: ${DATA_DIR}`);
    
    let grandTotalProcessed = 0;
    let grandTotalErrors = 0;
    
    // Process each exchange directory
    for (const [exchangeDir, exchangeName] of Object.entries(EXCHANGES)) {
        const exchangePath = path.join(DATA_DIR, exchangeDir);
        
        if (fs.existsSync(exchangePath)) {
            console.log(`\n📊 Processing ${exchangeDir} (${exchangeName})...`);
            
            const result = await processDirectory(exchangePath, exchangeName);
            grandTotalProcessed += result.processed;
            grandTotalErrors += result.errors;
            
            console.log(`✅ ${exchangeDir}: ${result.processed.toLocaleString()} records, ${result.errors} errors`);
        }
    }
    
    console.log(`\n🎉 Import Complete!`);
    console.log(`📊 Total records processed: ${grandTotalProcessed.toLocaleString()}`);
    console.log(`❌ Total errors: ${grandTotalErrors.toLocaleString()}`);
    
    // Update data freshness
    const updateStmt = db.prepare(`
        INSERT OR REPLACE INTO data_freshness 
        (symbol, last_updated, record_count, earliest_date, latest_date, error_count) 
        VALUES (?, datetime('now'), ?, ?, ?, ?)
    `);
    
    // Get summary for each symbol
    const summaryStmt = db.prepare(`
        SELECT symbol, COUNT(*) as count, MIN(date) as earliest, MAX(date) as latest
        FROM historical_prices 
        GROUP BY symbol
    `);
    
    summaryStmt.all([], (err, rows) => {
        if (err) {
            console.log(`❌ Error getting summary: ${err.message}`);
            return;
        }
        
        rows.forEach(row => {
            updateStmt.run([
                row.symbol,
                row.count,
                row.earliest,
                row.latest,
                0 // no errors
            ]);
        });
        
        updateStmt.finalize();
        summaryStmt.finalize();
        
        console.log(`\n💾 Database updated with ${rows.length} symbols`);
        db.close();
    });
}

// Run the import
importAllData().catch(console.error);
